### ISSUE TITLE
Impossible to download newest flight from LXNAV S8/S80/S10/S100 when more than 128 flights in logger

### ISSUE THREAD

**Issue by groundhogxc** [XCSoar/XCSoar#1237]
XCSoar versions having and not having the problem
-------------------------------------------------

All versions, probably

Expected behavior
-----------------

When trying to download flight from the logger,  the list should show all flights in the logger, or at least the most recent ones

Actual behavior
---------------

The list shows the 128 *oldest* flights, newer flights can not be downloaded.

Do you have any idea what may have caused this?
-----------------------------------------------

XCSoar puts an arbitrary limit for the download list at 128 entries

Do you have an idea how to solve the issue?
-------------------------------------------

Easiest: We just increase the number of entries to a value that is higher than the max number of stored flights (which is unfortunately unknown). I tested 512, and so far it works, but in a few years we might reach 512 and it might fail again. I did not find a way to artificially generate flights to test the limit without doing actual flights.

(See pull request for this quick solution)

Otherwise: Use a list without fixed length.

Better solution: Only download the *newest* N flights rather than the oldest. Not sure if the protocol allows that easily, I was not able to understand it in reasonably short time by reading the quite complicated download code.

**Comment by Scumi** [comment]
The LXNAV Data Port Specification says the following:

- You can get the log book size with `PLXVC,LOGBOOKSIZE,R`
- The device answers with `PLXVC,LOGBOOKSIZE,A,<int> `

I just checked very briefly, but the correct implementation is probably in `NanoLogger.cpp ` as the Nanos and the S-Varios share (almost) the same protocol.

**Comment by groundhogxc** [comment]
Yes, I did understand that NanoLogger.cpp is the relevant file, but there it just uses directly the "RecordedFlightList" structure with the hard coded length, instead of dynamically requesting the logbook size. So all LXNAV devices including the Nano should be affected, no? The logbook size seems to be ignored for all of them.

At least this driver does not crash in this situation like the IMI one did :-)

**Comment by MaxKellermann** [comment]
Oh, that data structure was designed when nobody could imagine that loggers could store so many files - pilots were happy they didn't have to manually delete flights - older loggers would just stop working if memory was full.

Enlarging the data structure would solve this issue, but listing so many flights would take a loooong time, so showing only the newest N flights (for an arbitrary value of N, currently 128) seems like a reasonable compromise, doesn't it?

**Comment by MaxKellermann** [comment]
Off-topic, but if you feel some piece of code is in need of more code comments, write a bug report - unfortunately, nobody reviews my contributions and grouches at me for writing more documentation (the same way I do it to others), and there are indeed some less-than-perfect corners in the XCSoar code base. I'll gladly clarify what is unclear.
